File: Python/Two Pointers/16. 3Sum Closest.py
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        res = [float('inf')]
        for i in range(len(nums)):
            self.twoSum(nums, i, target, res)
        return res[0] + target
        
    def twoSum(self, nums, index, target, res):
        left, right = index+1, len(nums)-1
        while left < right:
            s = nums[index] + nums[left] + nums[right]
            if abs(s-target) < abs(res[0]): res[0] = s - target
            if res[0] == 0:
                break
            elif s - target < 0:
                left += 1
            else:
                right -= 1        

File: C++/Two Pointers/K Sum Series/16. 3Sum Closest.cpp
#include <algorithm>
#include <cmath>

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        // residual
        int res = INT_MAX;
        for (int i = 0; i < nums.size(); i++) {
            twoSum(nums, i, target, res);
            if (res == 0) break;
        }
        return res + target;
    }
    
    void twoSum(vector<int>& nums, int index, int target, int& res) {
        int left = index+1, right = nums.size()-1;
        while (left < right) {
            int s = nums[index] + nums[left] + nums[right];
            if (abs(s-target) < abs(res)) res = s - target;
            if (s - target == 0) {
                break;
            }
            else if (s - target < 0) {
                left++;
            }
            else {
                right--;
            }
        }   
    }
};